ykrt/src/location.rs

```diff
@@ -97,88 +29,407 @@ impl State {
+            }
+        }
+    }
+}
+
+#[cfg(target_pointer_width = "64")]
+pub(crate) const STATE_TAG: usize = 0b111; // All of the other tag data must fit in this.
+#[cfg(target_pointer_width = "64")]
+pub(crate) const STATE_NUM_BITS: usize = 3;
+
+pub(crate) const STATE_IS_LOCKED: usize = 0b001;
+pub(crate) const STATE_IS_PARKED: usize = 0b010;
+pub(crate) const STATE_IS_COUNTING: usize = 0b100;
+
+pub(crate) const TOKEN_NORMAL: UnparkToken = UnparkToken(0);
+pub(crate) const TOKEN_HANDOFF: UnparkToken = UnparkToken(1);
+
+#[derive(PartialEq, Eq, Debug)]
+pub(crate) struct State<I> {
+    pub(crate) x: usize,
+    marker: PhantomData<I>,
+}
+
+impl<I> Copy for State<I> {}
+
+impl<I> Clone for State<I> {
+    fn clone(&self) -> State<I> {
+        *self
+    }
+}
+
+impl<I> State<I> {
+    /// Return a new `State` in the counting phase with a count 0.
+    pub(crate) fn new() -> Self {
+        State {
+            x: STATE_IS_COUNTING,
+            marker: PhantomData,
+        }
+    }
+
+    fn from_usize(x: usize) -> Self {
+        State {
+            x,
+            marker: PhantomData,
+        }
+    }
+
+    pub(crate) fn is_locked(&self) -> bool {
+        self.x & STATE_IS_LOCKED != 0
+    }
+
+    pub(crate) fn is_parked(&self) -> bool {
+        self.x & STATE_IS_PARKED != 0
+    }
+
+    /// Is the Location in the counting or a non-counting state?
+    pub(crate) fn is_counting(&self) -> bool {
+        self.x & STATE_IS_COUNTING != 0
+    }
+
+    /// If, and only if, the Location is in the counting state, return the current count.
+    pub(crate) fn count(&self) -> usize {
+        debug_assert!(self.is_counting());
+        debug_assert!(!self.is_locked());
+        self.x >> STATE_NUM_BITS
+    }
+
+    /// If this `State` is not counting, return its `HotLocation`. It is undefined behaviour to
+    /// call this function if this `State` is in the counting phase and/or if this `State` is not
+    /// locked.
+    pub(crate) unsafe fn hot_location(&self) -> &mut HotLocation<I> {
+        debug_assert!(!self.is_counting());
+        //debug_assert!(self.is_locked());
+        &mut *((self.x & !STATE_TAG) as *mut _)
+    }
+
+    /// Return a version of this `State` with the locked bit set.
+    pub(crate) fn with_lock(&self) -> State<I> {
+        State {
+            x: self.x | STATE_IS_LOCKED,
+            marker: PhantomData,
+        }
+    }
+
+    /// Return a version of this `State` with the locked bit unset.
+    pub(crate) fn with_unlock(&self) -> State<I> {
+        State {
+            x: self.x & !STATE_IS_LOCKED,
+            marker: PhantomData,
+        }
+    }
+
+    /// Return a version of this `State` with the parked bit set.
+    pub(crate) fn with_parked(&self) -> State<I> {
+        State {
+            x: self.x | STATE_IS_PARKED,
+            marker: PhantomData,
+        }
+    }
+
+    /// Return a version of this `State` with the parked bit unset.
+    pub(crate) fn with_unparked(&self) -> State<I> {
```

Comment:
Is `without_parked` a better name? The logic seems upside-down to me, but my head is a weird place. U p to you.

Reply:
Good question! I'm not super happy with the name -- let me chew on it.

Reply:
or merge `with_parked` and `with_unparked` into a `with_parked(val: u8)`?

Reply:
Hmm, honestly, I'm not sure. I'm starting to think that this might end up being part of the `State` rethink, and perhaps is best deferred until then.